### PR TITLE
Transform result to dict before returning from Job.result

### LINKED ISSUE #942: job.result() returns a string representation of dict rather than the dict itself
### What is the expected enhancement?

job.result() should translate the string into a dictionary for the user. We could probably catch serialization errors in the client and provide helpful error messages if the ones coming from `json` (or whatever) aren't informative enough.

### PR DESCRIPTION
**Pull request by caleb-johnson**
Fixes #942

### CODE REVIEW

File: client/quantum_serverless/core/job.py [hunk 1]
```diff
@@ -375,7 +375,11 @@ def result(self, wait=True, cadence=5, verbose=False):
                 time.sleep(cadence)
                 if verbose:
                     logging.info(".")
-        return self._job_client.result(self.job_id)
+        results = self._job_client.result(self.job_id)
+        if isinstance(results, str):
```
Comment (psschwei) [inline client/quantum_serverless/core/job.py:379]: I could be overthinking things here, but would it be better to handle this within the job client (i.e. [here](https://github.com/Qiskit-Extensions/quantum-serverless/blob/c2a5e9d6d5adde09eb20590e9163298ada593154/client/quantum_serverless/core/job.py#L286))?

Comment (caleb-johnson) [reply client/quantum_serverless/core/job.py:379]: It looks like the return is already loaded into a dictionary in that line, but when I use `save_result` inside my remote program, `job.result()` returns me a string representation of a dictionary

Comment (caleb-johnson) [reply client/quantum_serverless/core/job.py:379]: OK I see what's happening. I think I am intended to use the client to retrieve results, but I was using the `job` object itself. I guess it's appropriate to do the decoding on both levels?

Comment (caleb-johnson) [reply client/quantum_serverless/core/job.py:379]: @IceKhan13 why would the return object in the deleted line ever be a string? `job_client.result()` returns a decoded object [here](https://github.com/Qiskit-Extensions/quantum-serverless/blob/c2a5e9d6d5adde09eb20590e9163298ada593154/client/quantum_serverless/core/job.py#L286)

Comment (IceKhan13) [reply client/quantum_serverless/core/job.py:381]: Hmm, good question. It should be decoded object (dict). I do not know why it's returning string. Worth digging into it

File: client/quantum_serverless/core/job.py [hunk 2]
```diff
@@ -375,7 +375,18 @@ def result(self, wait=True, cadence=5, verbose=False):
...
                 if verbose:
                     logging.info(".")
-        return self._job_client.result(self.job_id)
+
+        # Retrieve the results. If they're string format, try to decode to a dictionary.
+        results = self._job_client.result(self.job_id)
+        if isinstance(results, str):
+            try:
+                results = json.loads(
+                    self._job_client.result(self.job_id), cls=QiskitObjectsDecoder
```
Comment (IceKhan13) [inline client/quantum_serverless/core/job.py:384]:
```suggestion
                    results, cls=QiskitObjectsDecoder
```

**Comment by IceKhan13** [comment]
@caleb-johnson can you update code to pass lints and we can merge this :)

**Comment by caleb-johnson** [comment]
> @caleb-johnson can you update code to pass lints and we can merge this :)

I can do that, but I am not really satisfied that this is the right fix. I do not understand why we need to decode the string here, when it is [already being done in the GatewayJobClient](https://github.com/Qiskit-Extensions/quantum-serverless/blob/97caeb6a09b116a2b5e58bb376f2eb93d084f567/client/quantum_serverless/core/job.py#L276).

Do you understand what is causing these JSON strings to make it through the `result` call?

**Comment by psschwei** [comment]
Something looks a bit off with the commits...

**Comment by caleb-johnson** [comment]
> Something looks a bit off with the commits...

*sigh* 

> Something looks a bit off with the commits...

Thanks, I cleaned it up. I made a mess by merging main into this branch :P

Going to take a closer look at this now that Aki solved the other problem with results

**Comment by IceKhan13** [comment]
@caleb-johnson thing is we have DB schema in following format

```
Model: Job
Fields:
- id: str
- status: str
- logs: str
- results: str
```

so when it is returned from DB json looks like this
```json
{
  "id": "42",
  "status": "SUCCEEDED",
  "results": "{'bla-bla': 'bla-bla'}"
}
```

therefore we need to do another decoding on `results` field.

It can be fixed by changing https://github.com/Qiskit-Extensions/quantum-serverless/blob/main/gateway/api/models.py#L94 to JsonField, probably. Or even better, by creating custom serializer which will decode results field before returning back to user https://github.com/Qiskit-Extensions/quantum-serverless/blob/main/gateway/api/serializers.py#L23

File: client/quantum_serverless/core/job.py [hunk 3]
```diff
@@ -375,7 +375,16 @@ def result(self, wait=True, cadence=5, verbose=False):
...
                     logging.info(".")
-        return self._job_client.result(self.job_id)
+
+        # Retrieve the results. If they're string format, try to decode to a dictionary.
+        results = self._job_client.result(self.job_id)
+        if isinstance(results, str):
+            try:
+                results = json.loads(results, cls=QiskitObjectsDecoder)
+            except json.JSONDecodeError:
+                pass
```
Comment (IceKhan13) [inline client/quantum_serverless/core/job.py:385]: can we log something here?

```
logger.warning("Error during results decoding. Details: %s", exception)
```

Comment (IceKhan13) [reply client/quantum_serverless/core/job.py:385]: and we are good to go :)

**Comment by caleb-johnson** [comment]
@IceKhan13 , as discussed, we can just merge this patch and follow up with a better fix in a following PR.

Added a warning when the string isn't interpretable by json, as requested.

Ready for final review/approval

**Review by IceKhan13 — APPROVED** [review]

**Pull request merged**